Add dialog exit functions

mildmojo opened the pull request:
Adds {exit} and {exitNow} functions to exit to a new room from
dialog (including conditionals). See comments for usage and
installation instructions.

seleb commented on exit-from-dialog.js:
```diff
@@ -0,0 +1,165 @@
+           the BOTTOM left tile is (0,0). If you'd like to use sprite coordinates,
+           add the word "sprite" as the fourth parameter to the exit function.
+
+  Installation:
```
make this bit consistent with others as mentioned in other PR

seleb commented on exit-from-dialog.js:
```diff
@@ -0,0 +1,165 @@
+  var queuedDialogExit = null;
+
+  // Hook into the game reset and make sure exit data gets cleared.
+  var _clearGameData = clearGameData;
```
`globals.clearGameData`

mildmojo replied:
Same note as on the other PR.

seleb commented on exit-from-dialog.js:
```diff
@@ -0,0 +1,165 @@
+  };
+
+  // Hook into the dialog finish event; if there was an {exit}, travel there now.
+  var _onExitDialog = onExitDialog;
```
`globals.onExitDialog`

seleb commented on exit-from-dialog.js:
```diff
@@ -0,0 +1,165 @@
+
+  // Implement the {exitAfter} dialog function. It saves the room name and
+  // destination X/Y coordinates so we can travel there after the dialog is over.
+  globals.exitAfterFunc = function(environment, parameters, onReturn) {
```
it'd probably be safer to inject the actual function code for the exits into the script instead of putting them in global scope and injecting references to them, e.g.
```js
var exitAfterFunc = function(){...};
...
_inject(
'var functionMap = new Map();',
'functionMap.set("exitAfter", ' + exitAfterFunc.toString() + ');'
);
```

mildmojo replied:
Injecting the source of the function inline won't give it access to functions/variables in my enclosure unless I expose them globally. Somethin's gotta be global! 🙂

seleb replied:
ah good point; you could inject the whole kit and caboodle but for simplicity's sake this makes sense then

seleb commented on exit-from-dialog.js:
```diff
@@ -0,0 +1,165 @@
+  // and X/Y coordinates right damn now.
+  globals.exitNowFunc = function(environment, parameters, onReturn) {
+    var exitParams = _getExitParams(parameters);
+    if (!exitParams) return;
```
brackets :p

seleb commented on exit-from-dialog.js:
```diff
@@ -0,0 +1,165 @@
+  Example: {exitAfter "FinalRoom,8,4"}
+           {exitNow "FinalRoom,8,11,sprite"}
+
+  WARNING: In exit coordinates, the TOP left tile is (0,0). In sprite coordinates,
```
should move this above the usage+examples, seeing "sprite" in those is confusing until you've seen this bit

seleb commented on exit-from-dialog.js:
```diff
@@ -0,0 +1,165 @@
+    var y = params[2];
+    var coordsType = (params[3] || 'exit').toLowerCase();
+    var useSpriteCoords = coordsType === 'sprite';
+    var roomId = names.room.get(roomName);
```
`globals.names`

seleb commented on exit-from-dialog.js:
```diff
@@ -0,0 +1,165 @@
+
+    // recreate the script and dialog objects so that they'll be
+    // referencing the code with injections instead of the original
+    scriptModule = new Script();
```
`globals.` on all of these

mildmojo commented:
I synced up the documentation format and cleaned the code up a bit. I also renamed the exit function name back to `{exit}` instead of `{exitAfter}` because it's the one you're probably going to use the most and probably has the most predictable behavior.
